### PR TITLE
Revised train-test splitter to support artist conditioning and distribution matching

### LINKED ISSUE #24: De-duping
Inspired by a recent tweet on deduping in cifar, I went and performed a similar analysis on our proposed train-test split.  We indeed have some dupes in our dataset, though it seems like they're relatively few.  This does already occur in our (current) train-test split though, so we'll need to do some cleanup to prevent that from happening.

### Procedure

1. Map all tracks to vggish quantized codewords, and flatten the time dimension.  This gives for train a 15000x1280 matrix; for test, 5000x1280.
2. Perform a nearest neighbor search from test to train using L1 distance (though any distance would work).
3. Look for any suspiciously small distances to nearest neighbors.

### Results

image

This plot shows the distribution of distances from test to 1st, 2nd, ..., 5th nearest neighbor from the training set.  As you can see from the fliers on column 1, there are a few suspiciously low values.  In fact, two of them are identically 0.  The offending culprits are

```
007954_184320 -> 120195_184320
074658_372480 -> 074328_372480
```
Checking the metadata fields, these appear to be exact duplicates with dirty metadata.

There's one more soft near-match, which has distance 9057 (median is about 54000):
```
021334_11520 -> 020898_11520
```
which again, by metadata, seems like a like match.  (The artist ids are identical, but the other metadata is garbage.)

Since in all of these cases, an artist-id filter using metadata would have caught this, I'm wondering if that's sufficient to fix it?  I didn't include this originally because I was under the impression that no artist had more than one track in the collection.  Apparently that assumption was incorrect.

I haven't done a similar analysis on the train-train data, but I suspect there are many more dupes than are caught here.

What do you recommend @ejhumphrey @simondurand ?  I could try to resolve this in the splitter, but I'm not sure how artist conditioning will play with stratified sampling -- in sklearn splitters, we can do either stratification or grouping, but not both simultaneously.  I'll take a look and see if we just got unlucky with this random split, but it'd be nice to have a proper solution in place.

### PR DESCRIPTION
**Pull request by bmcfee**
Fixes #24 

This PR overhauls the train-test split generator in three ways:

1. The metadata has been de-duped at the artist id level using a nearest-neighbor search in vggish feature space.  The process is described in #24 comment thread, with a link the corresponding notebook.  The process is deterministic and easy to reproduce.
2. Train-test splits are now artist-conditional, and can optionally use the de-dupe index produced by (1) above.  This effectively eliminates the most obvious cases of train-test leakage by (near) duplication.
3. Shuffle splits are now done with rejection sampling based on a probability ratio match.  For each proposed split, the distribution of positive and negative examples for each instrument cannot deviate by more than a constant factor from the full sample.

For step (3), I have checked in a generated train-test split with tolerance of 0.875 (closer to 1 meaning a tighter match to the full distribution).  I haven't been able to improve on this yet, and we're probably hitting the limits imposed by the data (e.g. the Bach effect mentioned in #24), but I think this ratio is pretty good.

Aside from CR, what I need from @ejhumphrey and/or @simondurand is instructions for where to put the dedupe index data.  Otherwise, I think this resolves all of the major issues known in the previous train-test split logic.

### CODE REVIEW

File: scripts/openmic_split.py [hunk 1]
```diff
@@ -34,8 +35,11 @@ def process_args(args):
                         help='Path to sparse-labels.csv',
                         type=str)
 
+    parser.add_argument('--dupes', dest='dupe_file', type=str,
+                        help='Path to track de-duplication index')
+
     parser.add_argument('-s', '--seed', dest='seed',
-                        default=20180821,
+                        default=20180903,
```
Comment (ejhumphrey) [inline scripts/openmic_split.py]: nitwarning: if we're going to use dates, maybe we can pull this from date-time so we won't have to keep bumping it? otherwise it's probably fine to leave it as some pseudo-arbitrary number, e.g. 12345?

File: scripts/openmic_split.py [hunk 2]
```diff
@@ -84,13 +104,55 @@ def load_label_matrix(metadata_file, label_file):
...
+    '''
+    min_prob, max_prob = sorted([prob_ratio, 1./prob_ratio])
+
+    all_dist_p = (label_matrix > 0).sum() / label_matrix.count()
+    all_dist_n = (label_matrix <= 0).sum() / label_matrix.count()
+
+    sub_dist_p = (label_matrix.loc[idx] > 0).sum() / label_matrix.loc[idx].count()
+    sub_dist_n = (label_matrix.loc[idx] <= 0).sum() / label_matrix.loc[idx].count()
+
+    return (np.all(min_prob * all_dist_p.values <= sub_dist_p.values) and
```
Comment (ejhumphrey) [inline scripts/openmic_split.py]: not sure if this is the best place to do so, but looking at this logic for the first time / without a lot of context, I'm having trouble unpacking the intuition here. Would an inline comment make sense? Or in the docstring?

File: scripts/openmic_split.py [hunk 3]
```diff
@@ -116,26 +178,46 @@ def make_partitions(metadata, labels, seed, num_splits, ratio):
...
-        if set(train_ser) & set(test_ser):
+    fold = 0
+    for artist_train_idx, artist_test_idx in tqdm(splitter.split(labels, labels)):
+        train_artists = artist_labels.index[artist_train_idx]
+        test_artists = artist_labels.index[artist_test_idx]
+
+        train_idx = sample_keys[sample_keys['artist_id'].isin(train_artists)]['sample_key'].sort_values()
+        test_idx = sample_keys[sample_keys['artist_id'].isin(test_artists)]['sample_key'].sort_values()
+
+        if set(train_idx) & set(test_idx):
```
Comment (ejhumphrey) [inline scripts/openmic_split.py]: is this taking the intersection? not familiar with this syntax (could have seen this as the union?)

**Comment by ejhumphrey** [comment]
LGTM for the most part, just a couple parts that could use a comment or two to help readability.

File: scripts/openmic_split.py (hunk 3 above)
Comment (bmcfee) [reply scripts/openmic_split.py]: Yup, `set(A) & set(B)` = `{x | x in A AND x in B}`.

File: scripts/openmic_split.py (hunk 1 above)
Comment (bmcfee) [reply scripts/openmic_split.py]: I was taking the date of writing as a pseudo-arbitrary number for the seed, but one that's unlikely to be randomly reproduced elsewhere.  It's a silly convention, and I for sure didn't need to change the seed here, but there's no harm in leaving it here.

Current date-time is exactly what I don't want, since that won't be fixed on subsequent runs.

File: scripts/openmic_split.py (hunk 2 above)
Comment (bmcfee) [reply scripts/openmic_split.py]: Yeah, a comment makes sense here.  I'll add it.

**Comment by bmcfee** [comment]
> high-level question: how big is the "dedupe index data"? is that just the dupe_file named here that falls out of the notebook?

Yep, exactly:
```
-rw-rw-r--   1 bmcfee bmcfee    378583 Sep  3 15:50 openmic-20k-dedupe-artistids.csv
```
20001 lines (header + dedupe id for each sample key).

**Comment by bmcfee** [comment]
> LGTM for the most part, just a couple parts that could use a comment or two to help readability.

I made the requested comments, and think we're good to merge.  Travis is currently failing on most platforms for reasons independent of this PR (dependency installation).

File: scripts/openmic_split.py [hunk 4]
```diff
@@ -84,13 +104,60 @@ def load_label_matrix(metadata_file, label_file):
...
+    min_prob, max_prob = sorted([prob_ratio, 1./prob_ratio])
+
+    all_dist_p = (label_matrix > 0).sum() / label_matrix.count()
+    all_dist_n = (label_matrix <= 0).sum() / label_matrix.count()
+
+    sub_dist_p = (label_matrix.loc[idx] > 0).sum() / label_matrix.loc[idx].count()
+    sub_dist_n = (label_matrix.loc[idx] <= 0).sum() / label_matrix.loc[idx].count()
+
+    # Make sure that for each class Y, we have the following conditions:
+    #   P[Y = 1 | x in sample] >= P[Y = 1] * min_prob   # positive examples are not too unlikely
```
Comment (ejhumphrey) [inline scripts/openmic_split.py]: 🙏

**Comment by ejhumphrey** [comment]
ah, okay, I wondered what the hell was going on with Travis, same thing was happening for the full label aggregation. We don't actually add any tests here, so I suppose it's fine to mail on through?

only thing that stands out now is the notebook that produced the de-duped index ... I know it's linked in a gist, that's linked in the PR ... but perhaps that's a little bit buried? maybe if we don't want to add the notebook directly to the repo, we could add the link to a description in the README.md? this is definitely non-blocking though, since we can add it anytime.

**Comment by bmcfee** [comment]
> We don't actually add any tests here, so I suppose it's fine to mail on through?

:+1: 



> this is definitely non-blocking though, since we can add it anytime.

:+1: yeah.  I think it makes sense to have it in the repo, but it's non-blocking.  Maybe we can just have a separate folder for supplemental materials / derived data?

**Comment by ejhumphrey** [comment]
travis is complaining about env issues and taking forever to run, this PR is fine...
